src/addon/branchmonitor.js

```diff
@@ -0,0 +1,301 @@
+                        if (newAddOn instanceof Error) {
+                            logger.error('Could not get addOn', newAddOn);
+                        } else {
+                            runningAddOnsNew.push({
+                                id: requiredAddOns[i],
+                                instance: newAddOn
+                            });
+                        }
+                    }
+                }
+
+                self.runningAddOns = runningAddOnsNew;
+                self.commitMessage = 'AddOns';
+
+                return updateRunningAddOns(commitData.commitObject);
+            })
+            .then(function () {
+                var persisted = core.persist(self.rootNode),
+                    persistedObjects = Object.keys(persisted.objects);
+                if (persistedObjects.length === 0) {
+                    logger.debug('No changes made by addOns');
+                } else {
+                    // This will create an update event with local:true, see top of function.
+                    logger.debug('Changes were made', self.commitMessage);
+                    logger.debug('Number of persisted objects', persistedObjects.length);
+                    project.makeCommit(branchName, [self.commitHash], persisted.rootHash,
+                        persisted.objects, self.commitMessage,
+                        function (err, result) {
+                            if (err) {
+                                logger.error('makeCommit', err);
+                            } else {
+                                logger.debug('makeCommit', result);
+                            }
+                        });
+                }
+                handlerCallback(null, self.stopRequested === false);
+            })
+            .catch(function (err) {
+                logger.error('Fatal error', err);
+                handlerCallback(err, false);
+            });
+    }
+
+    function branchStatusHandler(branchStatus /*, commitQueue, updateQueue*/) {
+        //TODO check how it should work
+        logger.debug('New branchStatus', branchStatus);
+    }
+
+    // API functions
+    /**
+     * Opens up its branch and registers the onUpdate function.
+     * @param [callback]
+     * @returns {Promise}
+     */
+    this.start = function (callback) {
+        if (self.startRequested === false) {
+            startDeferred = Q.defer();
+            self.startRequested = true;
+
+            storage.openBranch(project.projectId, branchName, onUpdate, branchStatusHandler,
+                function (err/*, latestCommitData*/) {
+                    if (err) {
+                        startDeferred.reject(err);
+                        return;
+                    }
+                    self.branchIsOpen = true;
+                    startDeferred.resolve();
+                }
+            );
+        }
+
+        return startDeferred.promise.nodeify(callback);
```

Comment:
Yes, that is happening all the time.
In addOnManager.monitorBranch (invoked on each socket room join) each invocation reuses this promise. It's a neat feature of the q library. Kevin suggested that for mongo connections in #482.